File: legacy-python/cursor_orch/repl.py
from __future__ import annotations

from pathlib import Path

from prompt_toolkit import PromptSession
from prompt_toolkit.completion import WordCompleter
from prompt_toolkit.history import FileHistory
from rich.console import Console

from cursor_orch.commands import (
    COMMANDS,
    cmd_help,
    cmd_prompt_set,
    cmd_run,
    prompt_set_command_text,
    setup_summary_lines,
    validate_model_value,
    validate_prompt_value,
)
from cursor_orch.config import OrchestratorConfig
from cursor_orch.session import Session

_NON_MUTATION = {"help", "config", "repos", "run"}


def _build_completer() -> WordCompleter:
    words = [f"/{k}" for k in COMMANDS]
    words.extend(["/exit", "/quit", "/clear"])
    return WordCompleter(words)


def _parse_input(raw: str) -> tuple[str, list[str]]:
    parts = raw[1:].split()
    if not parts:
        return "", []
    cmd = parts[0]
    args = parts[1:]
    if cmd == "repo" and args and args[0] == "remove":
        return "repo-remove", args[1:]
    return cmd, args


def _read_multiline(prompt_session: PromptSession, console: Console) -> str | None:
    lines: list[str] = []
    while True:
        try:
            line = prompt_session.prompt("... ")
        except KeyboardInterrupt:
            console.print("[red]Prompt input cancelled.[/red]")
            return None
        except EOFError:
            break
        if line == "":
            break
        lines.append(line)
    if not lines:
        return None
    return "\n".join(lines)


def _handle_run(session: Session, console: Console) -> OrchestratorConfig | None:
    result = cmd_run(session)
    if result is None:
        return None
    if "errors" in result:
        for err in result["errors"]:
            console.print(f"[red]Error:[/red] {err}")
        return None
    if "config" in result:
        session.clear_setup_state()
        session.save_session()
        return result["config"]
    return None


def _is_control(value: str, control: str) -> bool:
    return value.strip().lower() == control


def _run_guided_setup(
    session: Session,
    prompt_session: PromptSession[str],
    console: Console,
) -> OrchestratorConfig | None:
    console.print("Welcome to cursor-orch interactive setup.")
    console.print("I will ask only what is required for your first run.")
    console.print("Controls: type 'skip' (use default), 'back' (previous step), or 'exit' (cancel setup).")

    setup_state = session.setup_state()
    step = str(setup_state.get("step", "model"))
    if step not in {"model", "prompt", "confirm"}:
        step = "model"
    if step == "confirm" and not session.has_required_guided_values():
        step = "model"

    session.set_setup_state(active=True, step=step)
    session.save_session()

    while True:
        if step == "model":
            console.print("Step 1/2 - Model")
            try:
                raw = prompt_session.prompt("AI model [default: gpt-5]: ")
            except KeyboardInterrupt:
                console.print("Setup cancelled. Returning to REPL.")
                session.set_setup_state(active=True, step="model")
                session.save_session()
                return None
            except EOFError:
                console.print("Setup closed. Returning to REPL.")
                session.set_setup_state(active=True, step="model")
                session.save_session()
                return None
            text = raw.strip()
            if _is_control(text, "exit"):
                session.set_setup_state(active=True, step="model")
                session.save_session()
                return None
            if _is_control(text, "back"):
                console.print("Already at first step.")
                continue
            if text == "" or _is_control(text, "skip"):
                value = "gpt-5"
            else:
                error = validate_model_value(text)
                if error:
                    console.print(f"[red]{error}[/red]")
                    continue
                value = text
            session.set_model(value)
            session.set_setup_state(active=True, step="prompt")
            session.save_session()
            console.print(f"Model set to: {value}")
            console.print(f"Equivalent command: /model {value}")
            step = "prompt"
            continue

        if step == "prompt":
            console.print("Step 2/2 - Prompt")
            console.print("Enter orchestration prompt. Finish with an empty line.")
            console.print("Prompt (required):")
            lines: list[str] = []
            while True:
                try:
                    line = prompt_session.prompt("... ")
                except KeyboardInterrupt:
                    console.print("Setup cancelled. Returning to REPL.")
                    session.set_setup_state(active=True, step="prompt")
                    session.save_session()
                    return None
                except EOFError:
                    console.print("Setup closed. Returning to REPL.")
                    session.set_setup_state(active=True, step="prompt")
                    session.save_session()
                    return None
                if not lines:
                    stripped = line.strip()
                    if _is_control(stripped, "back"):
                        step = "model"
                        session.set_setup_state(active=True, step="model")
                        session.save_session()
                        break
                    if _is_control(stripped, "exit"):
                        session.set_setup_state(active=True, step="prompt")
                        session.save_session()
                        return None
                    if _is_control(stripped, "skip"):
                        console.print("Prompt is required for first run. Enter prompt text or type 'back'/'exit'.")
                        continue
                if line == "":
                    prompt_text = "\n".join(lines)
                    error = validate_prompt_value(prompt_text)
                    if error:
                        console.print(error)
                        lines = []
                        continue
                    session.set_prompt(prompt_text)
                    session.set_setup_state(active=True, step="confirm")
                    session.save_session()
                    console.print(f"Prompt captured ({len(prompt_text)} characters).")
                    console.print("Equivalent command: /prompt")
                    console.print(f"Equivalent command: {prompt_set_command_text(prompt_text)}")
                    step = "confirm"
                    break
                lines.append(line)
            continue

        if step == "confirm":
            console.print("Setup complete. Review before execution:")
            for line in setup_summary_lines(session):
                console.print(line)
            console.print("Next actions:")
            console.print("1) run   -> execute now")
            console.print("2) back  -> edit previous step")
            console.print("3) exit  -> cancel and return to REPL")
            while True:
                try:
                    action = prompt_session.prompt("Choose action [run/back/exit]: ").strip().lower()
                except KeyboardInterrupt:
                    console.print("Setup cancelled. Returning to REPL.")
                    session.set_setup_state(active=True, step="confirm")
                    session.save_session()
                    return None
                except EOFError:
                    console.print("Setup closed. Returning to REPL.")
                    session.set_setup_state(active=True, step="confirm")
                    session.save_session()
                    return None
                if action == "back":
                    step = "prompt"
                    session.set_setup_state(active=True, step="prompt")
                    session.save_session()
                    break
                if action == "exit":
                    session.set_setup_state(active=True, step="confirm")
                    session.save_session()
                    return None
                if action == "run":
                    console.print("Equivalent command: /run")
                    result = cmd_run(session)
                    if result is None:
                        console.print("Run blocked: unknown error")
                        console.print("Type 'back' to edit required input or 'exit' to return to REPL.")
                        continue
                    if "errors" in result:
                        console.print(f"Run blocked: {result['errors'][0]}")
                        console.print("Type 'back' to edit required input or 'exit' to return to REPL.")
                        continue
                    if "config" in result:
                        session.clear_setup_state()
                        session.save_session()
                        console.print("Setup complete. Launching run.")
                        return result["config"]
                else:
                    console.print("Invalid choice. Type run, back, or exit.")
            continue


def _dispatch(
    cmd: str,
    args: list[str],
    session: Session,
    console: Console,
) -> str | None:
    cmd_info = COMMANDS[cmd]
    if cmd == "help":
        return cmd_help()
    try:
        return cmd_info.handler(session, *args)
    except TypeError:
        return f"[red]Usage:[/red] {cmd_info.usage}"


def run_repl() -> OrchestratorConfig | None:
    session = Session()
    console = Console()

    history_dir = Path.home() / ".cursor-orch"
    history_dir.mkdir(parents=True, exist_ok=True)
    history_path = history_dir / "history"

    prompt_session: PromptSession[str] = PromptSession(
        history=FileHistory(str(history_path)),
        completer=_build_completer(),
    )

    console.print("[bold]cursor-orch v0.2.0[/bold]")
    console.print("[dim]Type /help for available commands.[/dim]")
    console.print("[dim]Next: complete guided setup (or set prompt via /prompt), then run /run.[/dim]")
    console.print("[dim]Before /run, ensure CURSOR_API_KEY and GH_TOKEN are set (copy .env.example to .env).[/dim]")

    resumed = session.load_session()
    if resumed:
        console.print("[green]Resumed previous session. Type /config to review.[/green]")

    direct_commands_entered = False
    setup_state = session.setup_state()
    setup_active = bool(setup_state.get("active", False))
    if setup_active:
        step = str(setup_state.get("step", "model"))
        console.print(f"[yellow]Resuming guided setup from step: {step}.[/yellow]")
    should_run_guided = session.should_resume_guided_setup() or (
        (not session.has_required_guided_values()) and not direct_commands_entered
    )
    if should_run_guided:
        config = _run_guided_setup(session, prompt_session, console)
        if config is not None:
            return config

    while True:
        try:
            text = prompt_session.prompt("> ").strip()
        except KeyboardInterrupt:
            continue
        except EOFError:
            break

        if not text:
            continue

        if not text.startswith("/"):
            console.print("Unknown input. Type /help for available commands.")
            continue

        direct_commands_entered = True

        cmd, args = _parse_input(text)
        if not cmd:
            continue

        if cmd in ("exit", "quit"):
            return None

        if cmd == "clear":
            console.clear()
            continue

        if cmd not in COMMANDS:
            console.print(f"[red]Unknown command:[/red] /{cmd}. Type /help for available commands.")
            continue

        if cmd == "prompt":
            console.print(COMMANDS["prompt"].handler(session))
            multiline_text = _read_multiline(prompt_session, console)
            if multiline_text is not None:
                result = cmd_prompt_set(session, multiline_text)
                console.print(result)
                session.save_session()
            continue

        if cmd == "run":
            config = _handle_run(session, console)
            if config is not None:
                return config
            continue

        output = _dispatch(cmd, args, session, console)
        if output:
            console.print(output)

        if cmd not in _NON_MUTATION:
            session.save_session()

    return None

File: src/repl.ts
import * as fs from "node:fs";
import * as path from "node:path";
import * as readline from "node:readline/promises";
import tty from "node:tty";
import type { Interface as ReadlineInterface } from "node:readline/promises";
import {
  COMMANDS,
  cmdHelp,
  cmdPromptSet,
  cmdRun,
  promptSetCommandText,
  setupSummaryLines,
  validateModelValue,
  validatePromptValue,
} from "./commands.js";
import type { OrchestratorConfig } from "./config/types.js";
import { Session } from "./session.js";
import { readReplLineTTY } from "./lib/repl/tty-line-editor.js";
import { readVersion } from "./version.js";
import { tui } from "./tui/style.js";

const NON_MUTATION = new Set(["help", "config", "repos", "run"]);

function parseInput(raw: string): [string, string[]] {
  const parts = raw.slice(1).trim().split(/\s+/);
  if (!parts.length || !parts[0]) {
    return ["", []];
  }
  const cmd = parts[0]!;
  const args = parts.slice(1);
  if (cmd === "repo" && args[0] === "remove") {
    return ["repo-remove", args.slice(1)];
  }
  return [cmd, args];
}

function isControl(value: string, control: string): boolean {
  return value.trim().toLowerCase() === control;
}

function isInteractiveTty(): boolean {
  return (
    tty.isatty(0) &&
    tty.isatty(1) &&
    typeof process.stdin.setRawMode === "function"
  );
}

type RlHolder = { rl: ReadlineInterface };

async function readLineOrEof(rl: ReadlineInterface, prompt: string): Promise<string | null> {
  return new Promise((resolve, reject) => {
    let settled = false;
    const finish = (action: () => void): void => {
      if (settled) {
        return;
      }
      settled = true;
      rl.off("close", onClose);
      action();
    };
    const onClose = (): void => {
      finish(() => {
        resolve(null);
      });
    };
    rl.once("close", onClose);
    void rl
      .question(prompt)
      .then((line) => finish(() => resolve(line)))
      .catch((err: unknown) => finish(() => reject(err)));
  });
}

async function runGuidedSetup(
  session: Session,
  holder: RlHolder,
  historyPath: string,
): Promise<OrchestratorConfig | null> {
  console.log("Welcome to cursor-orch interactive setup.");
  console.log("I will ask only what is required for your first run.");
  console.log("Controls: type 'skip' (use default), 'back' (previous step), or 'exit' (cancel setup).");

  let step = session.setupState().step;
  if (!["model", "prompt", "confirm"].includes(step)) {
    step = "model";
  }
  if (step === "confirm" && !session.hasRequiredGuidedValues()) {
    step = "model";
  }

  session.setSetupState({ active: true, step });
  session.saveSession();

  while (true) {
    if (step === "model") {
      console.log("Step 1/2 - Model");
      const raw = await readLineOrEof(holder.rl, "AI model [default: gpt-5]: ");
      if (raw === null) {
        return null;
      }
      const text = raw.trim();
      if (isControl(text, "exit")) {
        session.setSetupState({ active: true, step: "model" });
        session.saveSession();
        return null;
      }
      if (isControl(text, "back")) {
        console.log("Already at first step.");
        continue;
      }
      let value: string;
      if (text === "" || isControl(text, "skip")) {
        value = "gpt-5";
      } else {
        const err = validateModelValue(text);
        if (err) {
          console.log(tui.red(err));
          continue;
        }
        value = text;
      }
      session.setModel(value);
      session.setSetupState({ active: true, step: "prompt" });
      session.saveSession();
      console.log(`Model set to: ${value}`);
      console.log(`Equivalent command: /model ${value}`);
      step = "prompt";
      continue;
    }

    if (step === "prompt") {
      console.log("Step 2/2 - Prompt");
      const ttyPrompt = isInteractiveTty();
      if (ttyPrompt) {
        console.log(
          "Enter your orchestration prompt as plain text (TTY: Ctrl+J or Alt+Enter for newline, Enter to submit). Shift+Enter also sends Enter here and cannot be a newline unless your terminal maps it to a line feed. A leading / starts a command; here use plain text, or type back/exit.",
        );
      } else {
        console.log(
          "Enter your orchestration prompt as plain text at the prompt (single line in this environment). A leading / starts a command; here use plain text, or type back/exit.",
        );
      }
      let raw: string | null;
      if (ttyPrompt) {
        holder.rl.close();
        process.stdin.resume();
        raw = await readReplLineTTY(historyPath);
        holder.rl = readline.createInterface({
          input: process.stdin,
          output: process.stdout,
          terminal: true,
        });
      } else {
        raw = await readLineOrEof(holder.rl, "> ");
      }
      if (raw === null) {
        session.setSetupState({ active: true, step: "prompt" });
        session.saveSession();
        return null;
      }
      const stripped = raw.trim();
      if (isControl(stripped, "back")) {
        step = "model";
        session.setSetupState({ active: true, step: "model" });
        session.saveSession();
        continue;
      }
      if (isControl(stripped, "exit")) {
        session.setSetupState({ active: true, step: "prompt" });
        session.saveSession();
        return null;
      }
      if (isControl(stripped, "skip")) {
        console.log("Prompt is required for first run. Enter prompt text or type 'back'/'exit'.");
        continue;
      }
      if (!stripped) {
        console.log("Prompt is required for first run. Enter prompt text or type 'back'/'exit'.");
        continue;
      }
      if (stripped.startsWith("/")) {
        console.log(tui.yellow("At this step, enter the prompt as plain text (no leading /), or type back or exit."));
        continue;
      }
      const output = cmdPromptSet(session, raw);
      if (output) {
        console.log(output);
      }
      if (validatePromptValue(raw) !== null) {
        continue;
      }
      session.setSetupState({ active: true, step: "confirm" });
      session.saveSession();
      console.log(`Equivalent command: ${promptSetCommandText(session.config.prompt)}`);
      step = "confirm";
      continue;
    }

    if (step === "confirm") {
      console.log("Setup complete. Review before execution:");
      for (const line of setupSummaryLines(session)) {
        console.log(line);
      }
      console.log("Next actions:");
      console.log("1) run   -> execute now");
      console.log("2) back  -> edit previous step");
      console.log("3) exit  -> cancel and return to REPL");
      while (true) {
        const actionRaw = await readLineOrEof(holder.rl, "Choose action [run/back/exit]: ");
        if (actionRaw === null) {
          return null;
        }
        const action = actionRaw.trim().toLowerCase();
        if (action === "back") {
          step = "prompt";
          session.setSetupState({ active: true, step: "prompt" });
          session.saveSession();
          break;
        }
        if (action === "exit") {
          session.setSetupState({ active: true, step: "confirm" });
          session.saveSession();
          return null;
        }
        if (action === "run") {
          console.log("Equivalent command: /run");
          const result = cmdRun(session);
          if ("errors" in result) {
            console.log(`Run blocked: ${result.errors[0]}`);
            console.log("Type 'back' to edit required input or 'exit' to return to REPL.");
            continue;
          }
          session.clearSetupState();
          session.saveSession();
          console.log("Setup complete. Launching run.");
          return result.config;
        }
        console.log("Invalid choice. Type run, back, or exit.");
      }
      continue;
    }
  }
}

function dispatch(cmd: string, args: string[], session: Session): string | null {
  const cmdInfo = COMMANDS[cmd];
  if (!cmdInfo) {
    return null;
  }
  if (cmd === "help") {
    return cmdHelp();
  }
  try {
    switch (cmd) {
      case "save":
        return cmdInfo.handler(session, args[0]) as string;
      case "load":
        return cmdInfo.handler(session, args[0] ?? "") as string;
      case "name":
        return cmdInfo.handler(session, args[0] ?? "") as string;
      case "model":
        return cmdInfo.handler(session, args[0] ?? "") as string;
      case "repo":
        return cmdInfo.handler(session, args[0] ?? "", args[1] ?? "", args[2] ?? "main") as string;
      case "repo-remove":
        return cmdInfo.handler(session, args[0] ?? "") as string;
      case "prompt-set":
        return cmdInfo.handler(session, args.join(" ")) as string;
      case "branch-prefix":
        return cmdInfo.handler(session, args[0] ?? "") as string;
      case "auto-pr":
        return cmdInfo.handler(session, args[0]) as string;
      case "bootstrap-repo":
        return cmdInfo.handler(session, args[0] ?? "") as string;
      case "save":
        return cmdInfo.handler(session, args[0]) as string;
      case "load":
        return cmdInfo.handler(session, args[0] ?? "") as string;
      default:
        return (cmdInfo.handler as (s: Session) => string)(session);
    }
  } catch {
    return `Usage: ${cmdInfo.usage}`;
  }
}

export async function runRepl(): Promise<OrchestratorConfig | null> {
  const session = new Session();
  const historyPath = session.historyPath;
  fs.mkdirSync(path.dirname(historyPath), { recursive: true });

  const holder: RlHolder = {
    rl: readline.createInterface({
      input: process.stdin,
      output: process.stdout,
      terminal: true,
    }),
  };

  console.log(tui.bold(`cursor-orch v${readVersion()}`));
  if (isInteractiveTty()) {
    console.log(
      tui.dim(
        "TTY line editor: Ctrl+J (or Alt+Enter) inserts a newline; Enter submits. Shift+Enter submits too in integrated terminals.",
      ),
    );
  }
  console.log(tui.dim("Plain text at > sets the orchestration prompt; lines starting with / are commands. Type /help for the list."));
  console.log(tui.dim("Next: complete guided setup (or set the prompt at > or via /prompt-set), then run /run."));
  console.log(tui.dim("Before /run, ensure CURSOR_API_KEY and GH_TOKEN are set (copy .env.example to .env)."));

  const resumed = session.loadSession();
  if (resumed) {
    console.log(tui.green("Resumed previous session. Type /config to review."));
  }

  let directCommandsEntered = false;
  if (session.setupState().active) {
    console.log(tui.yellow(`Resuming guided setup from step: ${session.setupState().step}.`));
  }
  const shouldRunGuided =
    session.shouldResumeGuidedSetup() || (!session.hasRequiredGuidedValues() && !directCommandsEntered);
  if (shouldRunGuided) {
    const cfg = await runGuidedSetup(session, holder, historyPath);
    if (cfg) {
      holder.rl.close();
      return cfg;
    }
  }

  const useTtyEditor = isInteractiveTty();
  if (useTtyEditor) {
    holder.rl.close();
    process.stdin.resume();
  }

  const readMainLine = (): Promise<string | null> =>
    useTtyEditor ? readReplLineTTY(historyPath) : readLineOrEof(holder.rl, "> ");

  try {
    while (true) {
      const rawLine = await readMainLine();
      if (rawLine === null) {
        return null;
      }
      const text = rawLine.trim();
      if (!text) {
        continue;
      }
      if (text.startsWith("/") && rawLine.includes("\n")) {
        console.log(
          tui.yellow(
            "Slash commands must be on one line. Use one line for /commands, or enter a multi-line prompt without a leading /.",
          ),
        );
        continue;
      }
      if (!text.startsWith("/")) {
        const output = cmdPromptSet(session, rawLine);
        if (output) {
          console.log(output);
        }
        if (validatePromptValue(rawLine) === null) {
          session.saveSession();
        }
        continue;
      }
      directCommandsEntered = true;
      const [cmd, args] = parseInput(text);
      if (!cmd) {
        continue;
      }
      if (cmd === "exit" || cmd === "quit") {
        return null;
      }
      if (cmd === "clear") {
        console.clear();
        continue;
      }
      if (!COMMANDS[cmd]) {
        console.log(tui.red(`Unknown command: /${cmd}. Type /help for available commands.`));
        continue;
      }
      if (cmd === "run") {
        const result = cmdRun(session);
        if ("errors" in result) {
          for (const err of result.errors) {
            console.log(tui.red(`Error: ${err}`));
          }
          continue;
        }
        session.clearSetupState();
        session.saveSession();
        return result.config;
      }
      const output = dispatch(cmd, args, session);
      if (output) {
        console.log(output);
      }
      if (!NON_MUTATION.has(cmd)) {
        session.saveSession();
      }
    }
  } finally {
    if (!useTtyEditor) {
      holder.rl.close();
    }
  }
}
